src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -55,57 +61,196 @@ internal static partial class JsonWriterHelper
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x5C)); // Reverse Solidus \
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x60)); // Grave Access `
+
+            return mask;
+        }
+#endif
+
+        public static unsafe int NeedsEscaping(ReadOnlySpan<byte> value, JavaScriptEncoder encoder)
+        {
+            fixed (byte* ptr = value)
             {
-                if (NeedsEscaping(value[idx]))
+                int idx = 0;
+
+                if (encoder != null)
                 {
+                    idx = encoder.FindFirstCharacterToEncodeUtf8(value);
                     goto Return;
                 }
-            }
 
-            idx = -1; // all characters allowed
+#if BUILDING_INBOX_LIBRARY
+                if (Sse2.IsSupported)
+                {
+                    sbyte* startingAddress = (sbyte*)ptr;
+                    while (value.Length - 16 >= idx)
+                    {
+                        Debug.Assert(startingAddress >= ptr && startingAddress <= (ptr + value.Length - 16));
+
+                        // Load the next 16 bytes.
+                        Vector128<sbyte> sourceValue = Sse2.LoadVector128(startingAddress);
+
+                        // Check if any of the 16 bytes need to be escaped.
+                        Vector128<sbyte> mask = CreateEscapingMask(sourceValue);
+
+                        int index = Sse2.MoveMask(mask.AsByte());
+                        // If index == 0, that means none of the 16 bytes needed to be escaped.
+                        // TrailingZeroCount is relatively expensive, avoid it if possible.
+                        if (index != 0)
+                        {
+                            // Found at least one byte that needs to be escaped, figure out the index of
+                            // the first one found that needed to be escaped within the 16 bytes.
+                            idx += BitOperations.TrailingZeroCount(index | 0xFFFF0000);
+                            goto Return;
+                        }
+                        idx += 16;
+                        startingAddress += 16;
+                    }
+
+                    // Process the remaining characters.
+                    Debug.Assert(value.Length - idx < 16);
+                }
+#endif
+
+                for (; idx < value.Length; idx++)
+                {
+                    Debug.Assert((ptr + idx) <= (ptr + value.Length));
+                    if (NeedsEscaping(*(ptr + idx)))
+                    {
+                        goto Return;
+                    }
+                }
 
-        Return:
-            return idx;
+                idx = -1; // all characters allowed
+
+            Return:
+                return idx;
+            }
         }
 
         public static unsafe int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
```

Comment:
Reordering this to be effectively:

```
if ((value.Length < 16) || !Sse2.IsSupported)
{
    SoftwareFallback();
}
else
{
    Sse2Impl();
    SoftwareFallback(); // To Handle Trailing Bits
}
```

Is likely going to be more efficient (likely inlining the logic, rather than having method calls).

Reply:
`Loops` are predicted to be "taken" by default; so you will currently get a branch mispredict for the `length < 16` case and that regresses perf for that scenario more.

The JIT also, by default, emits `if (cond) { code1 } else { code2 }` blocks such that `code1` is the fallthrough case (predicted as taken by default on x86).

This means that you generally don't want to do `if ((value.Length > 16) && Sse2.IsSupported) { Sse2.Impl } SoftwareFallback();`; as that will also regress the small case more.

Ordering it as suggested above *should* help keep the small case fast and the branch mispredicts will be amortized for the longer cases.

Reply:
It doesn't seem to improve perf, and ends up hurting perf in the case when I call it with a variety of lengths within the benchmark.

Reply:
> and ends up hurting perf in the case when I call it with a variety of lengths within the benchmark.

That is likely a case of benchmarks training the branch predictor to expect 'x' and may not be representative of actual code.

Reply:
> That is likely a case of benchmarks training the branch predictor to expect 'x' and may not be representative of actual code.


That's why I chose random string lengths to avoid that.

```C#
    public unsafe class Test_EscapingComparison_Reorder
    {
        private string[] _sources;
        private JavaScriptEncoder _encoder;

        [GlobalSetup]
        public void Setup()
        {
            _sources = new string[1_000];
            var random = new Random(42);

            for (int j = 0; j < 1_000; j++)
            {
                int DataLength = random.Next(1, 10);

                var array = new char[DataLength];
                for (int i = 0; i < DataLength; i++)
                {
                    array[i] = (char)random.Next(97, 123);
                }
                _sources[j] = new string(array);
            }
            _encoder = null;
        }

        [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
        [Benchmark(Baseline = true)]
        public int NeedsEscapingNewFixed()
        {
            int result = 0;
            for (int i = 0; i < _sources.Length; i++)
            {
                result ^= NeedsEscaping_New_Fixed(_sources[i], _encoder);
            }
            return result;
        }

        [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
        [Benchmark]
        public int NeedsEscapingNewFixedReorder()
        {
            int result = 0;
            for (int i = 0; i < _sources.Length; i++)
            {
                result ^= NeedsEscaping_New_Fixed_Reorder(_sources[i], _encoder);
            }
            return result;
        }
      
       // ... implemenetations
}
```

I also tried with using the writer/serializer and didn't see improvement.